### PR TITLE
Differentiator

### LINKED ISSUE #18: Task 12
Rishi Leburu: I choose Task 12 (Differentiation Engine)

### PR DESCRIPTION
**Pull request by rleburu**
## Description
Implements the Differentiation Engine (Task 12). Provides two functions:
- `differentiate`: computes the derivative polynomial analytically, returning a PowerBasis and its coefficients. Supports higher-order derivatives via a `k` parameter. For non-PowerBasis inputs, converts via interpolation first.
- `evaluate_derivative`: evaluates the derivative numerically at given points using the Taylor table / undetermined coefficients approach. Supports `"centered"` (O(h^2)), `"forward"` (O(h)), and `"backward"` (O(h)) schemes, and arbitrary derivative orders via `k`.
- Adds `differentiate_coefficients()` to `PowerBasis` in `power.py`, which computes the derivative coefficients directly using the power rule. Both functions in the differentiator call this method.

## Linked Issue
Fixes #18 

## Implementation Checklist
- [ ] My class inherits from `PolynomialBasis` (if applicable).
- [ ] I have implemented all abstract methods (`evaluate_basis`, `fit`, etc.).
- [x] I have used `numpy` vectorization to avoid inefficient Python loops.
- [x] Documentation strings (docstrings) are provided for all public methods.

### 1. Mathematical Integrity
- [ ] **Convergence:** Does the implementation maintain expected accuracy? (Checked for Runge's phenomenon or stability).
- [ ] **Edge Cases:** Does it handle single points or non-equally spaced nodes correctly?
- [ ] **Domain Mapping:** Does the class correctly map the input domain to the basis domain (e.g., [-1, 1] for Chebyshev)?

### 2. API & Style
- [ ] **Method Signatures:** Methods match the Abstract Base Class exactly.
- [ ] **Naming:** Follows `CamelCase` for classes and `snake_case` for functions/variables.
- [ ] **Clean Code:** No "magic numbers"; constants are named and explained.

### 3. Testing Quality
- [ ] **Coverage:** Unit tests cover typical usage and edge cases.
- [ ] **Precision:** Tests use `np.testing.assert_allclose` instead of `==`.
- [ ] **Validation:** There is a test proving the basis can exactly reconstruct a polynomial of its own degree.

## Screenshots / Plots

### CODE REVIEW

**Comment by mtezzele** [comment]
Please rebase this PR

**Comment by rleburu** [comment]
Rebased onto the latest upstream/main! Additional Updates: added higher-order derivatives (`k` parameter), moved coefficient differentiation into `PowerBasis.differentiate_coefficients()`, and replaced centered differences with the Taylor table approach supporting centered/forward/backward schemes. 

Also, note: Affine tests depend on the fix in PR #23 where the y-nodes are not pulled back :).

File: src/numanalysislib/calculus/differentiator.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    Tuple[PowerBasis, np.ndarray]
+        A new PowerBasis of degree n-1 and its coefficients representing the derivative polynomial.
+    """
+    for _ in range(k):
+        if isinstance(basis, PowerBasis):
+            basis, coefficients = basis.differentiate_coefficients(coefficients)
+        else:
+            # For non-PowerBasis, fit to PowerBasis first, then differentiate
+            n = basis.degree
+            x_pts = np.linspace(basis.a, basis.b, n + 1)
```
Comment (AudreyBu) [inline src/numanalysislib/calculus/differentiator.py]: For non-PowerBasis, consider using its own nodes instead of equal-spaced nodes for all basis. e.g. Chebyshev polynomial, this may affect stability

File: src/numanalysislib/calculus/differentiator.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        Array of shape (n_dofs,) containing the basis coefficients.
+    k: int, optional
+        The order of the derivative to compute (default: 1).
+
+    Returns
+    -------
+    Tuple[PowerBasis, np.ndarray]
+        A new PowerBasis of degree n-1 and its coefficients representing the derivative polynomial.
+    """
+    for _ in range(k):
```
Comment (AudreyBu) [inline src/numanalysislib/calculus/differentiator.py:46]: consider case k > basis.degree

**Review by AudreyBu — COMMENTED** [review]
## Peer Review Checklist (To be completed by Reviewer)

### 1. Mathematical Integrity
- [x] **Convergence:** Does the implementation maintain expected accuracy? (Checked for Runge's phenomenon or stability).
- [ ] **Edge Cases:** Does it handle single points or non-equally spaced nodes correctly?
- [x] **Domain Mapping:** Does the class correctly map the input domain to the basis domain (e.g., [-1, 1] for Chebyshev)?

### 2. API & Style
- [x] **Method Signatures:** Methods match the Abstract Base Class exactly.
- [x] **Naming:** Follows `CamelCase` for classes and `snake_case` for functions/variables.
- [x] **Clean Code:** No "magic numbers"; constants are named and explained.

### 3. Testing Quality
- [x] **Coverage:** Unit tests cover typical usage and edge cases.
- [x] **Precision:** Tests use `np.testing.assert_allclose` instead of `==`.
- [n/a] **Validation:** There is a test proving the basis can exactly reconstruct a polynomial of its own degree.

## Screenshots / Plots

**Comment by mtezzele** [comment]
Please fix the failing tests

File: src/numanalysislib/calculus/differentiator.py (hunk 1 above)
Comment (rleburu) [reply src/numanalysislib/calculus/differentiator.py]: Done! We now use Chebyshev nodes for stability when interpolating.

File: src/numanalysislib/calculus/differentiator.py (hunk 2 above)
Comment (rleburu) [reply src/numanalysislib/calculus/differentiator.py:46]: Done! (Added to both functions, for efficiency)

**Comment by rleburu** [comment]
All tests should be passing now! Although we may need some additional test cases to be added once other PR's are merged.

File: src/numanalysislib/calculus/differentiator.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    x = np.asarray(x, dtype=float)
+
+    if scheme == "centered":
+        s = np.arange(-k, k + 1)
+    elif scheme == "forward":
+        s = np.arange(0, k + 1)
+    elif scheme == "backward":
+        s = np.arange(-k, 1)
+    else:
+        raise ValueError(f"Unknown scheme: {scheme}. Use 'centered', 'forward', or 'backward'.")
```
Comment (mtezzele) [inline src/numanalysislib/calculus/differentiator.py]: use the python `match-case` statement instead of `if-elif`

Comment (rleburu) [reply src/numanalysislib/calculus/differentiator.py]: Done!
